extensions/csv/src/main/java/io/deephaven/csv/densestorage/DenseStorageWriter.java

```diff
@@ -0,0 +1,122 @@
+     * is equal to this value)
+     */
+    private final QueueWriter.IntWriter controlWriter;
+    /**
+     * Byte sequences < DENSE_THRESHOLD are compactly stored here
+     */
+    private final QueueWriter.ByteWriter byteWriter;
+    /**
+     * Char sequences < DENSE_THRESHOLD are compactly stored here
+     */
+    private final QueueWriter.CharWriter charWriter;
+    /**
+     * Byte sequences >= DENSE_THRESHOLD are stored here
+     */
+    private final QueueWriter.ByteArrayWriter largeByteArrayWriter;
+    /**
+     * Strings >= DENSE_THRESHOLD are stored here
+     */
+    private final QueueWriter.CharArrayWriter largeCharArrayWriter;
+
+    public DenseStorageWriter() {
+        final int blobSize = DenseStorageConstants.LARGE_THRESHOLD * 1024;
+        this.controlWriter = new QueueWriter.IntWriter(100_000); // biggish
+        this.byteWriter = new QueueWriter.ByteWriter(blobSize);
+        this.charWriter = new QueueWriter.CharWriter(blobSize);
+        this.largeByteArrayWriter = new QueueWriter.ByteArrayWriter(100_000);
+        this.largeCharArrayWriter = new QueueWriter.CharArrayWriter(100_100);
+    }
+
+    public DenseStorageReader newReader() {
+        return new DenseStorageReader(
+                controlWriter.newReader(),
+                byteWriter.newReader(),
+                charWriter.newReader(),
+                largeByteArrayWriter.newReader(),
+                largeCharArrayWriter.newReader());
+
+    }
+
+    public void append(CharSlice cs) {
+        boolean fctrl = false;
+        boolean fbyte = false;
+        boolean fchar = false;
+        boolean flbyte = false;
+        boolean flchar = false;
+        final int size = cs.size();
+        final boolean isByteRepresentable = calcByteRepresentable(cs);
+        if (size >= DenseStorageConstants.LARGE_THRESHOLD) {
+            if (isByteRepresentable) {
+                final byte[] data = new byte[size];
+                cs.copyTo(data, 0);
+                flbyte = largeByteArrayWriter.addByteArray(data);
+                fctrl = controlWriter.addInt(DenseStorageConstants.LARGE_BYTE_SENTINEL);
+            } else {
+                final char[] data = new char[size];
+                cs.copyTo(data, 0);
+                flchar = largeCharArrayWriter.addCharArray(data);
+                fctrl = controlWriter.addInt(DenseStorageConstants.LARGE_STRING_SENTINEL);
+            }
+        } else {
+            // size < DenseStorageConstants.LARGE_THRESHOLD
+            if (isByteRepresentable) {
+                fbyte = byteWriter.addBytesFromCharSlice(cs);
+                fctrl = controlWriter.addInt(size);
+            } else {
+                fchar = charWriter.addChars(cs);
+                fctrl = controlWriter.addInt(-size);
+            }
+        }
+        // If any flushed, then flush them all. But try to be nice avoid double-flushing. This is annoying.
+        if (!fctrl && !fbyte && !fchar && !flbyte && !flchar) {
```

Comment:
I am confused as for why this is necessary.  It seems given the code below it is redundant.  So maybe it is an attempt at optimization?  I don't see how as an optimization attempt it could help tho.  So, I am confused :-).

Reply:
It's not redundant and it's not optimization.
"If any of them flushed, they all have to flush, except you don't have to flush if you just flushed"

Example: call them A, B, C, D, and E

if A, B, and E all just flushed, then we have to ask C and D to flush.

 Would it be clearer if I wrote it as:

~~~
if (fctrl  ||  fbyte  ||  fchar   ||  flbyte  || flchar)  {  // this line is different from before
       // this code is inside the braces now
        if (!fctrl) {
            controlWriter.flush();
        }
        if (!fbyte) {
            byteWriter.flush();
        }
        if (!fchar) {
            charWriter.flush();
        }
        if (!flbyte) {
            largeByteArrayWriter.flush();
        }
        if (!flbyte) {
            largeCharArrayWriter.flush();
        }
}
~~~

Reply:
Ok, yes, I see it now, and the logic as you wrote it originally does make sense.   But I do prefer your proposed alternative above, it makes it easier to read for me (and I got confused already so I would gladly take the alternative).

Thanks!

Reply:
Now simplified.